### PR TITLE
PLT-7982 - streaming events

### PR DESCRIPTION
**Pull request by willjgould**
Added a `WithAction` class, which executes arbitrary IO on events during indexing, a `Streamable` typeclass, which allows us to describe things we can stream to and from, and used these constructs to create a function `withStream`, that gives us a streaming indexer.

Pre-submit checklist:
- Branch
    - [x] Tests are provided (if possible)
    - [ ] Commit sequence broadly makes sense and have useful messages
    - [x] Important changes are reflected in changelog.d of the affected packages
    - [ ] Relevant tickets are mentioned in commit messages
- PR
    - [ ] (For external contributions) Corresponding issue exists and is linked in the description
    - [x] Targeting main unless this is a cherry-pick backport
    - [x] Self-reviewed the diff
    - [x] Useful pull request description
    - [ ] If relevant, reference the ADR in the PR and reference the PR in the ADR
    - [x] Reviewer requested

### CODE REVIEW

File: marconi-core/src/Marconi/Core/Transformer/WithStream/TChan.hs [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+   . (MonadIO m)
+  => TChan r
+  -> Stream (Of r) m ()
+streamFromTChan chan = repeatM . liftIO . atomically $ readTChan chan
+
+-- | A smart constructor for @WithAction@
+withStream
+  :: (Show event, Show (Point event))
+  => (Timed (Point event) event -> r)
+  -> TChan r
```
[the remark this answers is not in the document]
Comment (willjgould) [reply marconi-core/src/Marconi/Core/Transformer/WithStream/TChan.hs:29]: You're right, I'll switch to a `TBQueue`.

File: marconi-core/test/Marconi/CoreSpec.hs [hunk 2]
```diff
@@ -1636,6 +1706,74 @@ withRollbackFailureTest =
...
+            [ Tasty.testProperty "TBQueue streaming works" withStreamTBQueue
+            ]
+        , Tasty.testGroup
+            "withStreamSocket"
+            [ Tasty.testProperty "Socket streaming works" withStreamSocket
+            ]
+        ]
+    ]
+
+withStreamTBQueue :: Property
```
Comment (koslambrou) [inline marconi-core/test/Marconi/CoreSpec.hs:1728]: Property tests should start with `prop`. Also add a comment for each explaining _what_ your testing exactly.

**Review by koslambrou — COMMENTED** [review]
After taking a quick look, I got a few comments:

* for any new constructs and modules, like `WithAction`, `TBQueue` and `Socket`, I would really to see some top level module documentation explaining what capability the interface provides, and when it should be used, along with maybe some examples.
* why is the `Socket` module necessary? Will we actually need it?
* is there an issue with the TBQueue stream where if there is no consuming client, then the indexer simply blocks once the queue is full?

In general, the idea seems good. Once we _actually_ use it for the LedgerEvents, we can probably identify more clearly if that design works or not.

**Comment by willjgould** [comment]
@koslambrou 

> for any new constructs and modules, like WithAction, TBQueue and Socket, I would really to see some top level module documentation explaining what capability the interface provides, and when it should be used, along with maybe some examples.

Planning on it, just hadn't quite got round to it yet.

> why is the Socket module necessary? Will we actually need it?

I think we should chat about this tomorrow. I'm not sure, is the answer, but maybe.

> is there an issue with the TBQueue stream where if there is no consuming client, then the indexer simply blocks once the queue is full?

Good question, there probably is. I'll double check tomorrow and figure out a solution.

**Comment by berewt** [comment]
>> is there an issue with the TBQueue stream where if there is no consuming client, then the indexer simply blocks once the queue is full?

> Good question, there probably is. I'll double check tomorrow and figure out a solution.

It will. There isn't any workaround, it's the user's responsibility. Other solutions are: to let the memory leak or lose events. It looks worse to me.

File: marconi-core/src/Marconi/Core/Transformer/WithAction.hs [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+{- |
+  A transformer that fans events into an arbitrary action.
+
+ The configuration is a function @Timed (Point event) event -> IO ()@
+
+ See Marconi.Core.Transformer.WithStream.Socket and Marconi.Core.Transformer.WithStream.TBQueue
+ for example usages.
+
+ Once we are close enough to the tip of the chain, `WithCatchup` deactivate itself and pass directly
+ the blocks to the indexer.
```
Comment (koslambrou) [inline marconi-core/src/Marconi/Core/Transformer/WithAction.hs:13]: I don't understand this comment. What does that have to do with `WithAction`?

File: marconi-core/src/Marconi/Core/Transformer/WithAction.hs [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+ When you use several workers and coordinators, you may want to put the catchup on the final
+ indexers and not on the coordinators to improve performances.
```
Comment (koslambrou) [inline marconi-core/src/Marconi/Core/Transformer/WithAction.hs:16]: Same comment here.

File: marconi-core/src/Marconi/Core/Transformer/WithStream/Streamable.hs [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+import Network.Socket.ByteString.Lazy (sendAll)
+import Streaming (Of, Stream)
+import Streaming.Prelude (repeatM, yield)
+
+-- | Something that can have mapped events streamed to or from it
+class Streamable a r where
+  streamFrom :: a -> Stream (Of r) IO ()
+  streamTo :: (Timed (Point event) event -> r) -> a -> Timed (Point event) event -> IO ()
+
+{- | A streaming implementation using @Socket@. Here's an example of how
```
Comment (koslambrou) [inline marconi-core/src/Marconi/Core/Transformer/WithStream/Streamable.hs:21]: If that's already implemented in the test module, I would just point the user to that test code in the test module.

File: marconi-core/src/Marconi/Core/Transformer/WithStream/Streamable.hs [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+  streamTo mapping sock = sendAll sock . encode . mapping
+
+{- | An experimental streaming implementation using @TBQueue@.
+
+    IMPORTANT: Note that this will block when used with @withStream@ if you have slow or
+               non-existent consumers!
+
+    Here's an example of how to use it, in the form of a property test:
+
+    propWithStreamTBQueue :: Property
```
Comment (koslambrou) [inline marconi-core/src/Marconi/Core/Transformer/WithStream/Streamable.hs:98]: Same comment here as above.

**Review by koslambrou — APPROVED** [review]
LGTM. Minor comments.

File: marconi-core/src/Marconi/Core/Transformer/WithAction.hs (hunk 3 above)
Comment (willjgould) [reply marconi-core/src/Marconi/Core/Transformer/WithAction.hs:13]: Ah my bad, I copy and pasted the structure of `WithCatchup`'s description and must have forgotten to remove the rest!

**Pull request merged**